File: compiler/bench/kruskal.c
// expected: 4458
#include <stdio.h>
#define NODES 500
#define EDGES 500

int parent[NODES];
int rnk[NODES];
int edges[EDGES][3];  // from, to, weight

int find(int x) {
    while (parent[x] != x) x = parent[x];
    return x;
}

int unite(int x, int y) {
    int rx = find(x), ry = find(y);
    if (rx == ry) return 0;
    if (rnk[rx] < rnk[ry]) {
        parent[rx] = ry;
    } else if (rnk[rx] > rnk[ry]) {
        parent[ry] = rx;
    } else {
        parent[ry] = rx;
        rnk[rx]++;
    }
    return 1;
}

void init_uf() {
    for (int i = 0; i < NODES; i++) {
        parent[i] = i;
        rnk[i] = 0;
    }
}

void init_edges() {
    for (int i = 0; i < EDGES; i++) {
        edges[i][0] = i % (NODES - 1);
        edges[i][1] = i % (NODES - 1) + 1;
        edges[i][2] = (i % 17) + 1;
    }
}

void sort_edges() {
    for (int i = 1; i < EDGES; i++) {
        for (int j = 0; j < i; j++) {
            if (edges[j][2] > edges[j+1][2]) {
                int t0 = edges[j][0], t1 = edges[j][1], t2 = edges[j][2];
                edges[j][0] = edges[j+1][0];
                edges[j][1] = edges[j+1][1];
                edges[j][2] = edges[j+1][2];
                edges[j+1][0] = t0;
                edges[j+1][1] = t1;
                edges[j+1][2] = t2;
            }
        }
    }
}

int kruskal() {
    init_uf();
    int count = 0, weight = 0;
    for (int i = 0; i < EDGES && count < NODES - 1; i++) {
        if (unite(edges[i][0], edges[i][1])) {
            count++;
            weight += edges[i][2];
        }
    }
    return weight;
}

int main() {
    init_edges();
    sort_edges();
    printf("%d\n", kruskal());
    return 0;
}

File: compiler/bench/kruskal.fs
\ expected: 4458
\ Kruskal MST, 500 edges, sum of MST edge weights

500 constant NODES
500 constant EDGES
create parent NODES cells allot
create rank NODES cells allot
create edges EDGES 3 * cells allot  \ from, to, weight

: find ( x -- root )
  begin
    dup cells parent + @
    2dup <> while
    nip
  repeat
  drop ;

: union ( x y -- united? )
  find swap find
  2dup = if 2drop 0 exit then
  over cells rank + @
  over cells rank + @
  2dup < if
    2drop swap cells parent + !
  else
    2dup > if
      2drop cells parent + !
    else
      drop cells parent + !
      dup cells rank + @ 1+ swap cells rank + !
    then
  then
  1 ;

: init-uf ( -- )
  NODES 0 do
    i i cells parent + !
    0 i cells rank + !
  loop ;

: edge! ( from to w idx -- )
  3 * cells edges +
  dup rot swap 2 cells + !
  swap over 1 cells + !
  swap swap ! ;

: edge@ ( idx -- from to w )
  3 * cells edges +
  dup @
  over 1 cells + @
  rot 2 cells + @ ;

: init-edges ( -- )
  EDGES 0 do
    i NODES 1- mod
    i NODES 1- mod 1+
    i 17 mod 1+
    i edge!
  loop ;

: sort-edges ( -- )  \ bubble sort by weight
  EDGES 1 do
    i 0 do
      i edge@ drop nip
      i 1+ edge@ drop nip
      > if
        i edge@
        i 1+ edge@
        i edge!
        i 1+ edge!
      then
    loop
  loop ;

: kruskal ( -- mst-weight )
  init-uf
  0 0  \ count weight
  EDGES 0 do
    over NODES 1- >= if leave then
    i edge@
    rot drop
    union if
      swap 1+ swap
      i edge@ nip nip +
    then
  loop
  nip ;

: main init-edges sort-edges kruskal . cr ;
